### PR TITLE
fix(`cast`): reset `env.tx.caller` for impersonated txs

### LINKED ISSUE #9541: `cast run` fails on transactions that used an impersonated address
### Component

Cast

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (59f354c 2024-12-11T00:29:25.224313000Z)

### What command(s) is the bug in?

cast run

### Operating System

macOS (Apple Silicon)

### Describe the bug

Start anvil and send a transaction that uses address impersonation:

```
anvil --auto-impersonate --fork-url https://sepolia.base.org

cast send 0xF38aA5909D89F5d98fCeA857e708F6a6033f6CF8 0x60fe47b1000000000000000000000000000000000000000000000000000000000000000c --from 0x041563c07028Fc89106788185763Fc73028e8511 --unlocked
```

Call `cast run` with the resulting tx hash:

```
cast run <tx_hash>
Executing previous transactions from the block.
Error: EVM error; transaction validation error: nonce 1 too high, expected 0

Context:
- Error #0: EVM error
- Error #1: transaction validation error: nonce 1 too high, expected 0
- Error #2: nonce 1 too high, expected 0
```

Re-running `cast run` with trace logging reveals that the wrong `from` address was used, which explains why the call failed with a nonce error:

```
...
Executing previous transactions from the block.
2024-12-11T20:19:31.759135Z TRACE cast::cmd::run: executing call transaction tx=0xd35acd0109b1731ba2c95c08517d70ed34bc0985d3986c6b29b8d87663bfb5d8 to=0xf38aa5909d89f5d98fcea857e708f6a6033f6cf8
2024-12-11T20:19:31.759482Z TRACE transact:inspect: foundry_evm_core::backend: setting caller account acc=0xea9d06135a94773aaa59d25ffc62d2bd68fced70
2024-12-11T20:19:31.759498Z TRACE transact:inspect: foundry_evm_core::backend: allow cheatcode access account=0xea9d06135a94773aaa59d25ffc62d2bd68fced70
2024-12-11T20:19:31.759504Z TRACE transact:inspect: foundry_evm_core::backend: setting spec ID spec_id=CANCUN
2024-12-11T20:19:31.759510Z TRACE transact:inspect: foundry_evm_core::backend: setting test account acc=0xf38aa5909d89f5d98fcea857e708f6a6033f6cf8
2024-12-11T20:19:31.759515Z TRACE transact:inspect: foundry_evm_core::backend: add persistent account account=0xf38aa5909d89f5d98fcea857e708f6a6033f6cf8
2024-12-11T20:19:31.759527Z TRACE transact:inspect: foundry_evm_core::backend: allow cheatcode access account=0xf38aa5909d89f5d98fcea857e708f6a6033f6cf8
2024-12-11T20:19:31.760975Z TRACE transact:inspect: sharedbackend: request basic address=0xea9d06135A94773aAa59d25fFC62D2bd68fceD70
2024-12-11T20:19:31.761035Z TRACE backendhandler: received request basic address=0xea9d06135a94773aaa59d25ffc62d2bd68fced70
```

My best guess is that something tried to recover the signer from the transaction, which doesn't work for impersonated transactions because they have a fake signature.

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Closes #9541 

## Solution

Reset `env.tx.caller` to `tx.from` for impersonated txs

### CODE REVIEW

File: crates/cast/bin/cmd/run.rs [hunk 1]
```diff
@@ -212,6 +212,19 @@ impl RunArgs {
 
                     configure_tx_env(&mut env, &tx.inner);
 
+                    let is_impersonated = match tx.inner.inner {
```
Comment (grandizzy) [inline crates/cast/bin/cmd/run.rs:215]: I think can be simplified as
```rust
                    if let alloy_network::AnyTxEnvelope::Ethereum(ref envelope) = tx.inner.inner {
                        // If the transaction is impersonated, we need to set the caller to the from
                        // address Ref: https://github.com/foundry-rs/foundry/issues/9541
                        if tx_is_impersonated(envelope.signature(), envelope.ty()) {
                            env.tx.caller = tx.from;
                        }
                    }
```

**Review by grandizzy — COMMENTED** [review]
lgtm, minor nit, pls check

**Review by zerosnacks — APPROVED** [review]
lgtm

File: crates/cast/bin/cmd/run.rs [hunk 2]
```diff
@@ -251,6 +257,12 @@ impl RunArgs {
 
             configure_tx_env(&mut env, &tx.inner);
 
+            if is_impersonated_tx(&tx.inner.inner) {
```
Comment (DaniPopes) [inline crates/cast/bin/cmd/run.rs:260]: should this be moved to `configure_tx_env`?

**Pull request merged**